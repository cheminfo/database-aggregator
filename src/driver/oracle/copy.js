'use strict';

const model = require('../../mongo/model');
const sqlUtil = require('../../util/sql');

const common = require('./common');

module.exports = function (options) {
  common.checkOptions(['connectString', 'query', 'mode'], options);

  switch (options.mode) {
    case 'dateid':
      return dateid(options);
    default:
      throw new Error(`invalid mode: ${options.mode}`);
  }
};

async function dateid(options) {
  const collection = options.collection;
  const Model = model.getSource(collection);

  // connect to Oracle and Mongo
  const oracleConn = await common.connect(options);

  // search for latest document in Mongo
  const latest = await Model.findOne()
    .sort('-date')
    .exec();
  let query = options.query;

  query = `SELECT * FROM (\n${query}\n) inner_table`;

  if (latest) {
    // >= because many rows can have the same date
    query = `${query}\nWHERE moddate >= ${sqlUtil.formatTimestamp(
      latest.date
    )}`;
  }

  // This clause is very important for incremental updates
  query = `${query}\nORDER BY moddate ASC`;

  await common.copyEntries(oracleConn, query, collection);
  await oracleConn.release();
}
