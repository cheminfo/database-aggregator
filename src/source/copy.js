'use strict';

const model = require('../mongo/model');
const mongodbConnect = require('../mongo/connection');

const copyEntries = require('./copyEntries');
const getDriver = require('./getDriver');

async function copy(config) {
  const driver = getDriver(config.driver);

  const collection = config.collection;
  if (typeof collection !== 'string' || collection === '') {
    throw new TypeError('collection must be a string');
  }
  const Model = model.getSource(collection);

  await mongodbConnect();

  const latest = await Model.findOne()
    .sort('-date')
    .exec();

  await driver.getData(config, (data) => copyEntries(data, config), {
    latestDate: latest && latest.date
  });
}

module.exports = copy;
