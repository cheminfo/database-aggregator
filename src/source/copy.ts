'use strict';

import { getSource } from '../mongo/model';
import { connect } from '../mongo/connection';
import { ISourceConfigElement } from '../types';

import { copyEntries } from './copyEntries';
import { getDriver } from './getDriver';

export async function copy(config: ISourceConfigElement) {
  const driver = getDriver(config.driver);

  const collection = config.collection;
  if (typeof collection !== 'string' || collection === '') {
    throw new TypeError('collection must be a string');
  }
  const Model = getSource(collection);

  await connect();

  const latest = await Model.findOne()
    .sort('-date')
    .exec();

  await driver.getData(config, data => copyEntries(data, config), {
    latestDate: latest && latest.date
  });
}
