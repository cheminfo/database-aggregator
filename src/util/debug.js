'use strict';

const Debug = require('debug');

var error = Debug('aggregate:error');
var warn = Debug('aggregate:warn');
var debug = Debug('aggregate:debug');
var trace = Debug('aggregate:trace');

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

module.exports = function (prefix) {
  const func = (message) => debug(`(${prefix}) ${message}`);
  func.error = (message) => error(`(${prefix}) ${message}`);
  func.warn = (message) => warn(`(${prefix}) ${message}`);
  func.debug = func;
  func.trace = (message) => trace(`(${prefix}) ${message}`);
  return func;
};
