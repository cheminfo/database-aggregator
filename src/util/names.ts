const AGGREGATION = 'aggregation';
const SOURCE = 'source';
const META = 'meta';

const REMOVE_ID = 'source_remove_';
const COPY_ID = 'source_copy_';
const COPY_MISSING_ID = 'source_copy_missing_ids_';

export function getModelName(prefix: string, name: string) {
  return `${prefix}_${name}`;
}

export function getSourceName(name: string) {
  return getModelName(SOURCE, name);
}

export function getMetaModelName(name: string) {
  return getModelName(META, name);
}

export function getSourceModelName(name: string) {
  return getModelName(SOURCE, name);
}

export function getAggregationModelName(name: string) {
  return getModelName(AGGREGATION, name);
}

export function getCopyTaskId(collection: string) {
  return COPY_ID + collection;
}

export function getRemoveTaskId(collection: string) {
  return REMOVE_ID + collection;
}

export function getCopyMissingIdTaskId(collection: string) {
  return COPY_MISSING_ID + collection;
}
