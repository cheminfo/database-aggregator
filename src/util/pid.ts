/* eslint-disable no-process-exit */

const path = require("path");
const fs = require("fs");

const isRunning = require("is-running");

const pidFile = path.join(__dirname, "../../source.pid");

export function start() {
  try {
    const pid = fs.readFileSync(pidFile, "utf8");
    if (isRunning(pid)) {
      console.log(`process is running (pid: ${pid})`);
      process.exit(0);
    }
    console.log("pid file is here but process died");
    fs.unlinkSync(pidFile);
  } catch (e) {
    if (e.code === "ENOENT") {
      console.log("pid file does not exist yet");
    } else {
      throw e;
    }
  }
  const currentPid = process.pid;
  fs.writeFileSync(pidFile, String(currentPid));
}

export function stop(code?: number) {
  code = code || 0;
  fs.unlinkSync(pidFile);
  process.exit(code);
}
