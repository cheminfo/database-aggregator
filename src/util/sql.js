'use strict';

// http://stackoverflow.com/a/5133807
function twoDigits(d) {
  if (d >= 0 && d < 10) return `0${d.toString()}`;
  if (d > -10 && d < 0) return `-0${(-1 * d).toString()}`;
  return d.toString();
}

// exports.formatDate = function (date) {
//    return `TO_DATE('${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())}', 'YYYY-MM-DD')`;
// };

// exports.formatDateTime = function (date) {
//    return `TO_DATE('${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())} ${twoDigits(date.getUTCHours())}:${twoDigits(date.getUTCMinutes())}:${twoDigits(date.getUTCSeconds())}', 'YYYY-MM-DD HH24:MI:SS')`;
// };

exports.formatTimestamp = function (date) {
  return `(TIMESTAMP '${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())} ${twoDigits(date.getUTCHours())}:${twoDigits(date.getUTCMinutes())}:${twoDigits(date.getUTCSeconds())} UTC' AT TIME ZONE SESSIONTIMEZONE)`;
};
