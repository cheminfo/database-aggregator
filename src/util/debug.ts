import * as Debug from "debug";

let error = Debug("aggregator:error");
let warn = Debug("aggregator:warn");
let debug = Debug("aggregator:debug");
let trace = Debug("aggregator:trace");

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

export function debugUtil(prefix: string) {
  return {
    error: (message: string) => error(`(${prefix}) ${message}`),
    warn: (message: string) => warn(`(${prefix}) ${message}`),
    debug: (message: string) => debug(`(${prefix}) ${message}`),
    trace: (message: string) => trace(`(${prefix}) ${message}`),
  };
}
