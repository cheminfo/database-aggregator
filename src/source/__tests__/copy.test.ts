import { connect, disconnect } from '../../../test/mongoSetup';
import { clean, getCollection, getDriverPath } from '../../../test/util';
import { copy } from '../copy';

beforeAll(connect);
afterAll(disconnect);

describe('source copy', () => {
  it('should copy the data', async () => {
    const collection = getCollection('source_test');
    const config = {
      driver: getDriverPath('source-copy'),
      driverConfig: {
        driverValue: 2
      },
      collection: 'test'
    };

    await copy(config);
    await expect(collection.countDocuments()).resolves.toBe(1);
    let data = await collection.find().toArray();
    expect(clean(data)).toMatchSnapshot();

    await copy(config);
    await expect(collection.countDocuments()).resolves.toBe(3);
    data = await collection.find().toArray();
    expect(clean(data)).toMatchSnapshot();

    await copy(config);
    await expect(collection.countDocuments()).resolves.toBe(4);
    data = await collection.find().toArray();
    expect(clean(data)).toMatchSnapshot();

    await copy(config);
    await expect(collection.countDocuments()).resolves.toBe(4);
    const dataAfter = await collection.find().toArray();
    expect(clean(dataAfter)).toEqual(data);
  });

  it('should allow copy with big id', async () => {
    const collection = getCollection('source_test_big');
    const config = {
      driver: getDriverPath('source-copy-big'),
      collection: 'test_big'
    };

    await copy(config);
    const data = await collection.find().toArray();
    expect(clean(data)).toMatchSnapshot();
  });
});
