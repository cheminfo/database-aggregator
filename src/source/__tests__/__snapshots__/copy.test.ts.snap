// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source copy should allow copy with big id 1`] = `
Array [
  Object {
    "__v": 0,
    "commonID": "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
    "data": Object {
      "x": 42,
    },
    "date": 1970-01-01T00:00:00.000Z,
    "id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    "sequentialID": 1,
  },
]
`;

exports[`source copy should copy the data 1`] = `
Array [
  Object {
    "__v": 0,
    "commonID": "test1",
    "data": Object {
      "driverValue": 2,
      "x": 1,
      "y": 2,
    },
    "date": 1970-01-01T00:00:01.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
]
`;

exports[`source copy should copy the data 2`] = `
Array [
  Object {
    "__v": 0,
    "commonID": "test1",
    "data": Object {
      "x": 2,
      "y": 3,
    },
    "date": 1970-01-01T00:00:01.500Z,
    "id": "test1",
    "sequentialID": 2,
  },
  Object {
    "__v": 0,
    "commonID": "test2",
    "data": Object {
      "driverValue": 2,
      "x": 4,
      "y": 5,
    },
    "date": 1970-01-01T00:00:01.600Z,
    "id": "test2",
    "sequentialID": 3,
  },
  Object {
    "__v": 0,
    "commonID": "test4",
    "data": Object {
      "x": 8,
      "y": 9,
    },
    "date": 1970-01-01T00:00:01.600Z,
    "id": "test4",
    "sequentialID": 4,
  },
]
`;

exports[`source copy should copy the data 3`] = `
Array [
  Object {
    "__v": 0,
    "commonID": "test1",
    "data": Object {
      "x": 2,
      "y": 3,
    },
    "date": 1970-01-01T00:00:01.500Z,
    "id": "test1",
    "sequentialID": 2,
  },
  Object {
    "__v": 0,
    "commonID": "test2",
    "data": Object {
      "driverValue": 2,
      "x": 4,
      "y": 5,
    },
    "date": 1970-01-01T00:00:01.600Z,
    "id": "test2",
    "sequentialID": 3,
  },
  Object {
    "__v": 0,
    "commonID": "test4",
    "data": Object {
      "x": 8,
      "y": 9,
    },
    "date": 1970-01-01T00:00:01.600Z,
    "id": "test4",
    "sequentialID": 4,
  },
  Object {
    "__v": 0,
    "commonID": "test3",
    "data": Object {
      "x": 6,
      "y": 7,
    },
    "date": 1970-01-01T00:00:02.000Z,
    "id": "test3",
    "sequentialID": 5,
  },
]
`;
